Modelica/Electrical/Analog/Lines/Functions/lineZmatrix.mo

```diff
@@ -0,0 +1,116 @@
+  input Modelica.Units.SI.ResistancePerUnitLength R1[n] "Conductors lineic resistance";
+  input Real k_s[n]=fill(0.7,n) "Ratio of equivalent shell radius to actual radius";
+  //  in case of cylindric conductor this is equal to exp(-mu_r/4)=exp(-0.25)
+  input Modelica.Units.SI.Resistivity rho=100 "Ground resistivity";
+  input Modelica.Units.SI.Frequency f=50 "Frequency";
+  output Modelica.Units.SI.ResistancePerUnitLength Rcomp[div(n * (n + 1), 2)] "Compact resistance matrix";
+  output Modelica.Units.SI.ReactancePerUnitLength Xcomp[div(n * (n + 1), 2)] "Compact reactance matrix";
+  output Modelica.Units.SI.InductancePerUnitLength Lcomp[div(n * (n + 1), 2)] "Compact inductance)";
+protected
+  Modelica.Units.SI.Distance D "Generic larger distance";
+  Modelica.Units.SI.Distance d "Generic smaller distance";
+  Real Theta "Theta angle for Carson's formulas";
+  Real L "generic inductance";
+  Real P "P coefficient from Carson's original paper";
+  Real Q "Q coefficient from Carson's original paper";
+
+  Complex Z[n, n] "Computed matrix (ohm/m)";
+
+  final parameter Real w=2*pi*f;
+  parameter Real a0=sqrt(w*mu_0/rho) "multiplies D in a-parameter from EMTPs Theory Book";
+  // a0 is written, equivalently, as 4*pi*sqrt(5)*1e-4*sqrt(f/rho) in EMTP Theory book
+  //(reference available in Electrical.Analog.References)
+  //
+  //  a0*D=a; a is the same as "r" in Carson's paper
+  // where D=2h for self impedance, D=Dik for mutual impedance
+  Real a;
+  Integer k;
+algorithm
+//self impedances with ground return:
+  for i in 1:n loop
+    a:=a0*2*y[i];
+    assert(a<0.25,"parameter a="+String(a)+
+                  " >0.25 will result in inadequate precision of Carson's formulas",
+                  AssertionLevel.warning);
+    P := pi/8 - sqrt(2)/6*a + a^2/16*(0.672784 + log(2/a));
+    Q := (-0.03860784) + 0.5*log(2/a) + sqrt(2)/6*a-pi/64*a*a;
+    L := mu_0/(2*pi)*log(2*y[i]/(k_s[i]*r[i])) "L when soil is pure conductive";
+    Z[i, i] := R1[i] + 4*w*P/1.e7 + j*w*(L + 4*Q/1.e7); //Carson (25) and (30)
+  end for;
+//mutual impedances with ground return:
+  for i in 1:n loop
+    for jj in 1:i - 1 loop
+      Theta := atan(abs((x[i] - x[jj])/(y[i] + y[jj])));
+      d := sqrt((x[i] - x[jj])^2 + (y[i] - y[jj])^2); //distanza fra conduttore i e jj
+      D := sqrt((x[i] - x[jj])^2 + (y[i] + y[jj])^2);  //distanza fra conduttore i e immagine di jj
+      a:=a0*D;
+      P := pi/8 - sqrt(2)/6*a*cos(Theta) + a^2/16*(0.672784
+         + log(2/a))*cos(2*Theta)+a^2/16*Theta*sin(2*Theta);
+      Q := (-0.03860784) + 0.5*log(2/a) + sqrt(2)/6*a*cos(Theta)-pi/64*a*a*cos(2*Theta);
+      L := mu_0/(2*pi)*log(D/d) "L when soil is pure conductive";
+      Z[i, jj] := 4*w*P/1.e7 + j*w*(L + 4*Q/1.e7); //Carson (26) and (31)
+    end for;
+  end for;
+  k := 0;
+  for i in 1:n loop
+    for j in i:n loop
+      k := k + 1;
+      Rcomp[k] := Z[j, i].re;
+      Xcomp[k] := Z[j, i].im;
+      Lcomp[k] := Xcomp[k]/w;
+    end for;
```

Comment:
Even if Dymola doesn't report any conflict here, `j` is an imaginary unit (see import above). So e.g. `jj` could be used instead.

Reply:
There isn't any conflict, since for-loops introduce their own lexical scope.
However, it is a bit confusing  - as using the wrong `j` might silently give incorrect result.

This is a general issue for electrical models.

However, using `jj` isn't really better than `j` - as it looks confusing, and there's still the risk that someone writes `j` (meaning `jj`). How about using `row` (or `r`) and` column` (or `c)` instead of `i` and `j`?

Reply:
> There isn't any conflict, since for-loops introduce their own lexical scope.

Was not aware of it, sorry!

Using `row` / `column` or similar sound even better. Will do so in the future for my implementations.

Reply:
Nice discussion about the index names.

I believe that MSL code is to be corrent, efficient, compliant to Modelica.UsersGuide.Conventions. For the rest we should leave programmers the opportunity to choose their way to code. 
The same concept and formulas can be coded in infinite ways. I think that _a code that is universally agreed upon_ is just something that does not exist.  Every programmer have their own sensitivity and feelings.
It appears to me that the choice of the indexes of loops can be left to the programmers. I fact for a reader that prefers i, j there can be another which prefers k, l, another r, c, another row, col, etc.
I personally like i, j and also i, jj. They do not break any style rule, do not create inefficiencies, they just do their job in a  way that is clear enough. Maybe I decide, for homogeneity to switch everywhere to (i,j) or (i,jj), but this does not impact on the quality of the code.

Reply:
Sure. 
If for code review you indend also discussing the choice of names iteration variables in for loops I have no choice but to accept.
But.. may I say that I don't agree on this particular way of intending code review?

As you've seen I've already accepted (and appreciated) dozens of other comments and suggestions from you. Do you mean that authors must accept every comment without having to say a word?

Reply:
> Do you mean that authors must accept every comment without having to say a word?

Sure not! But developers who follow and contribute the MSL over years or even decades (not being myself) surely consider far more aspects when reviewing a code.

But let us concentrate on this particular case. There is a good reason to use iteration variable `jj` in
```
Z[i, jj] := 4*w*P/1.e7 + j*w*(L + 4*Q/1.e7);
```
So why not to follow this way further when referring the Z matrix index again?
```
Rcomp[k] := Z[j, i].re;
```

---
Btw: looking into EMTP theory book - the usage of "i" and "j"  is even more confusing. They use "i" for both el. current and index, and "j" for indexing and the imaginary unit. And to confuse the reader even more, the upright and italic variants of the two letters are mixed in the text and equations.

Reply:
> 
> But developers who follow and contribute the MSL over years or even decades (not being myself) surely consider far more aspects when reviewing a code.
> 

I'm 63 years old and I've been writing code during the last 35: Fortran, then C, then C++, then Matlab. then Modelica ;-)

> But let us concentrate on this particular case. There is a good reason to use iteration variable `jj` in
> 
> ```
> Z[i, jj] := 4*w*P/1.e7 + j*w*(L + 4*Q/1.e7);
> ```
> 
> So why not to follow this way further when referring the Z matrix index again?

Ok, let us go to "i" and "jj".

> 
> Btw: looking into EMTP theory book - the usage of "i" and "j" is confusing. They use "i" for both el. current and index, and "j" for indexing and the imaginary unit. And to confuse the reader even more, the upright and italic variants of the two letters are mixed in the text and equations.

Good that you took part of your time in checking Theory book.

People who worked on EMTP and EMTP theory book, during the seventies and eighties of the last century, were very brilliant and gifted, but clarity was not their main gift.
The situation did worsen when Theory book, originally written by hand, was typed. Typing introduced several typos, and also plain mistakes. I should have somewhere the scan of the manuscript pages for the part regarding lines, what I used (along with the analysis ot the original paper from Carson) to create my formulas in LineZmatrix and LineCmatrix, the correctness of which I now feel very confident... obviously taking into account that they are built around Carson's hypotheses, and that I truncated the infinite series after a few terms (with error checking and warning whenever the results is expected to be inaccurate).